components/service/glean/src/test/java/mozilla/components/service/glean/storages/EventsStorageEngineTest.kt

```diff
@@ -324,4 +339,122 @@ class EventsStorageEngineTest {
+
+        resetGlean(getContextWithMockedInfo(), Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ))
+
+        val event = EventMetricType<SomeExtraKeys>(
+            disabled = false,
+            category = "telemetry",
+            name = "test_event",
+            lifetime = Lifetime.Ping,
+            sendInPings = listOf("store1")
+        )
+
+        event.record(extra = mapOf(SomeExtraKeys.SomeExtra to "bar"))
+        assertEquals(1, event.testGetValue().size)
+
+        // Clear the in-memory storage only to mock being loaded in a fresh process
+        EventsStorageEngine.eventStores.clear()
+        resetGlean(
+            getContextWithMockedInfo(),
+            Glean.configuration.copy(
+                serverEndpoint = "http://" + server.hostName + ":" + server.port,
+                logPings = true
+            ),
+            clearStores = false
+        )
+
+        // Trigger worker task to upload the pings in the background
+        triggerWorkManager()
+
+        val request = server.takeRequest(20L, TimeUnit.SECONDS)
+        assertEquals("POST", request.method)
+        val applicationId = "mozilla-components-service-glean"
+        assert(
+            request.path.startsWith("/submit/$applicationId/store1/${Glean.SCHEMA_VERSION}/")
+        )
+        val pingJsonData = request.body.readUtf8()
+        val pingJson = JSONObject(pingJsonData)
+        checkPingSchema(pingJson)
+        assertNotNull(pingJson.opt("events"))
+        assertEquals(
+            1,
+            pingJson.getJSONArray("events").length()
+        )
+    }
+
+    @Test
+    fun `handle truncated events on disk`() {
+        val server = MockWebServer()
+        server.enqueue(MockResponse().setBody("OK"))
+
+        resetGlean(getContextWithMockedInfo(), Glean.configuration.copy(
+            serverEndpoint = "http://" + server.hostName + ":" + server.port,
+            logPings = true
+        ))
+
+        val event = EventMetricType<ExtraKeys>(
+            disabled = false,
+            category = "telemetry",
+            name = "test_event",
+            lifetime = Lifetime.Ping,
+            sendInPings = listOf("store1")
+        )
+
+        // Record the event in the store, without providing optional arguments.
+        event.record(mapOf(ExtraKeys.Key1 to "bar"))
+        EventsStorageEngine.testWaitForWrites()
+        // Add a couple of truncated events to disk. One is still valid JSON, the other isn't.
+        // These event should be skipped, all others intact.
+        EventsStorageEngine.writeEventToDisk("store1", "[500]")
+        EventsStorageEngine.testWaitForWrites()
+        EventsStorageEngine.writeEventToDisk("store1", "[500, \"foo")
+        EventsStorageEngine.testWaitForWrites()
+        event.record(mapOf(ExtraKeys.Key1 to "baz"))
+        EventsStorageEngine.testWaitForWrites()
```

Comment:
These additional `testWaitForWrites` is so these lines are written to the events file in order.  The failure in this test was due to checking the events in a different order in which they ended up in the file.

On this test, I'm pretty confident that these changes should help.